src/Compilers/CSharp/Portable/Symbols/OverriddenOrHiddenMembersHelpers.cs

```diff
@@ -930,33 +945,38 @@ internal static bool RequiresExplicitOverride(this MethodSymbol method)
                 {
                     return false;
                 }
+
                 // We can ignore newslot, since we checked IsOverride.
-                // We can ignore interface implementation changes since the method is already metadata virtual (since override).
-                // TODO: do we want to add more sophisticated handling for the case where there are multiple runtime-overridden methods?
-                MethodSymbol runtimeOverriddenMethod = method.GetFirstRuntimeOverriddenMethodIgnoringNewSlot(ignoreInterfaceImplementationChanges: true);
-                return runtimeOverriddenMethod is null ||
-                    csharpOverriddenMethod != runtimeOverriddenMethod &&
-                    method.IsAccessor() != runtimeOverriddenMethod.IsAccessor();
+                MethodSymbol runtimeOverriddenMethod = method.GetFirstRuntimeOverriddenMethodIgnoringNewSlot(out bool wasAmbiguous);
+                return wasAmbiguous
+                    // See https://github.com/dotnet/roslyn/issues/45453
+                    ? method.ContainingAssembly.RuntimeSupportsCovariantReturnsOfClasses
+                    : csharpOverriddenMethod != runtimeOverriddenMethod;
```

Comment:
>csharpOverriddenMethod != runtimeOverriddenMethod [](start = 22, length = 49)

Should we always return true if this condition is met and look at ambiguity otherwise?

Reply:
If the suggestion would lead to an incorrect behavior, it would be good to back the fact with a test demonstrating the fact.

---
In reply to: [446321935](https://github.com/dotnet/roslyn/pull/44025#discussion_r446321935) [](ancestors = 446321935)

Reply:
The test demonstrating that your suggestion would be incorrect is shown in [https://github.com/dotnet/runtime/issues/38119](https://github.com/dotnet/runtime/issues/38119) and, now that the runtime is fixed, is tested by the runtime team's unit tests.

I don't think it makes sense to add tests for a hypothetical buggy version of the compiler.

---
In reply to: [446358005](https://github.com/dotnet/roslyn/pull/44025#discussion_r446358005) [](ancestors = 446358005,446321935)

Reply:
> The test demonstrating that your suggestion would be incorrect is shown in [https://github.com/dotnet/runtime/issues/38119](https://github.com/dotnet/runtime/issues/38119) and, now that the runtime is fixed, is tested by the runtime team's unit tests.

Could you please elaborate what exactly would be wrong and how? 

> I don't think it makes sense to add tests for a hypothetical buggy version of the compiler.

It is always good to have test that prevents a regression in the future. Is there an existing test that would fail if the behavior of the function were to change as suggested? What would be the nature of the failure? 




---
In reply to: [446413877](https://github.com/dotnet/roslyn/pull/44025#discussion_r446413877) [](ancestors = 446413877,446358005,446321935)

Reply:
See `TestAmbiguousOverridesRefOut` in `CodeGenOverridingAndHiding` as changed in iteration 33. Depending on the order of the declarations, the override would change from being correct to incorrect or vice versa (both tested).

---
In reply to: [446416546](https://github.com/dotnet/roslyn/pull/44025#discussion_r446416546) [](ancestors = 446416546,446413877,446358005,446321935)

Reply:
> Adding the methodimpl entries changes what overridden method the runtime binds the override to. Since we want to remain compatible with previous behavior (right or wrong) for buggy runtimes, we do not want to change what we produce.

> See TestAmbiguousOverridesRefOut in CodeGenOverridingAndHiding as changed in iteration 33. Depending on the order of the declarations, the override would change from being correct to incorrect or vice versa (both tested).

I am just trying to understand why the code is written this way here. Could you, please, help me out with that?

At first, the logic feels wrong because we might not emit a MethodImpl for true covariant return case with ambiguous runtime override. However, one might argue that in the case when the runtime doesn't support covariant returns, we would report an error elsewhere and wouldn't even try to emit. Does this match the line of thinking you use? If so, consider capturing this in a comment.

On the other hand, could this negatively affect retargeting scenario when covariant returns involved? It looks like we would fail to communicate the presence of the covariant return override to the KnownOverriddenClassMethod above, potentially negatively affecting resolution of what the method overrides in that new universe. Is this possible? It might be worth adding some tests around this scenario. We might want to use a slightly different logic in this helper specifically for the purpose of RetargetingMethodSymbol.ExplicitlyOverriddenClassMethod.


---
In reply to: [446418702](https://github.com/dotnet/roslyn/pull/44025#discussion_r446418702) [](ancestors = 446418702,446416546,446413877,446358005,446321935)

Reply:
It looks like this helper also affects behavior of IsMetadataNewSlot API for source and retargeting scenario. Given all that, I would be more comfortable if the logic here was adjusted to accurately detect if we are dealing with covariant return situation and always return true for that case, regardless of  runtime capability, returning anything else feels incorrect for this scenario. For remaining cases, use logic based on the runtime capability. The current logic seems short and clever, but I am not comfortable with its possible inaccuracy in some scenarios. 

---
In reply to: [446444174](https://github.com/dotnet/roslyn/pull/44025#discussion_r446444174) [](ancestors = 446444174,446418702,446416546,446413877,446358005,446321935)

Reply:
> I am just trying to understand why the code is written this way here. Could you, please, help me out with that?

The compat council's decision means that we do not emit methodimpl in the ambiguous scenarios unless supported by the runtime. That is the *true* branch of the conditional. The false branch of the conditional simply checks if the runtime's unambiguous override candidate is the same as the language one. Methodimpl works on all runtimes for that scenario as far as I know.

> At first, the logic feels wrong because we might not emit a MethodImpl for true covariant return case with ambiguous runtime override. 

We only support covariant returns when targeting a platform that supports covariant returns. I don't think that needs additional documentation.

> On the other hand, could this negatively affect retargeting scenario when covariant returns involved?

This code is only when emitting the source method, and for the metadata properties of a source method. I don't think it can affect the properties of a retargeting method (which mirror the same metadata properties that would arise on the underlying method if emitted).
I don't think we intend to support retargeting a method that uses covariant returns to a runtime that does not support them. I don't think any small adjustment here could make that work.


---
In reply to: [446444174](https://github.com/dotnet/roslyn/pull/44025#discussion_r446444174) [](ancestors = 446444174,446418702,446416546,446413877,446358005,446321935)

Reply:
> The compat council's decision means that we do not emit methodimpl in the ambiguous scenarios unless supported by the runtime. That is the true branch of the conditional. The false branch of the conditional simply checks if the runtime's unambiguous override candidate is the same as the language one. Methodimpl works on all runtimes for that scenario as far as I know. 

That shouldn't affect the covariant return case because there is no prior behavior that we need to maintain and covariant returns always supposed to use MethodImpl. That is the design of the feature. Whether the runtime would understand it or not is orthogonal, in my opinion.

> This code is only used when emitting the source method

That is not the case, I just gave two examples. I would prefer us to implement the letter of the spec about covariant returns rather than hoping that this shortcut isn't going to have an adverse effect somewhere in the compiler today or in the future. 

> I don't think it can affect the properties of a retargeting method (which mirror the same metadata properties that would arise on the underlying method if emitted).

That would have to be proven by tests that observe that there is no behavior difference between source and retargeting references. Alternatively, we could align behavior of this method with Covariant Returns feature specification. I.e. This method should always return true for covariant return scenario, regardless of runtime capability. 

> I don't think any small adjustment here could make that work.

I am sorry, but I would have to disagree with that.

I can be convinced if you can come up with a test scenario that would demonstrate an incorrect behavior if the suggested changes were implemented.


---
In reply to: [446450749](https://github.com/dotnet/roslyn/pull/44025#discussion_r446450749) [](ancestors = 446450749,446444174,446418702,446416546,446413877,446358005,446321935)

Reply:
This code is not used for retargeting symbols. Retargeting symbols mirror the behavior of the PE symbols that would be generated for the underlying source symbols. Neither scenario "works" on platforms that do not support covariant returns.

> > This code is only used when emitting the source method

> That is not the case, I just gave two examples.

I don't know what examples you gave that suggests this code is used for anything other than a source symbol.

> > I don't think it can affect the properties of a retargeting method (which mirror the same metadata properties that would arise on the underlying method if emitted).

> That would have to be proven by tests that observe that there is no behavior difference between source and retargeting references. 

Retargeting symbols do not have any runtime behavior as they are compile-time artifacts.  Perhaps this code from `RetargetingMethodSymbol`'s base class should help:

```csharp
        internal override bool IsMetadataVirtual(bool ignoreInterfaceImplementationChanges = false)
        {
            return UnderlyingMethod.IsMetadataVirtual(ignoreInterfaceImplementationChanges);
        }

        internal override bool IsMetadataNewSlot(bool ignoreInterfaceImplementationChanges = false)
        {
            return UnderlyingMethod.IsMetadataNewSlot(ignoreInterfaceImplementationChanges);
        }
```

> I can be convinced if you can come up with a test scenario that would demonstrate an incorrect behavior if the suggested changes were implemented.

I have already done so. See TestAmbiguousOverridesRefOut in CodeGenOverridingAndHiding as changed in iteration 33. Depending on the order of the declarations, the override would change from being correct to incorrect or vice versa (both tested).

In any case, the compat council explicitly decided that we should not produce methodimpl entries for the ambiguous scenario on runtimes that are not known to have the fix to [https://github.com/dotnet/runtime/issues/38119](https://github.com/dotnet/runtime/issues/38119). The compat council's decision is documented in [https://github.com/dotnet/roslyn/issues/45453](https://github.com/dotnet/roslyn/issues/45453). I know you disagree with it, but it is simply not possible for me to implement your preferred approach and the compat council's decision at the same time, since they differ.


---
In reply to: [446453378](https://github.com/dotnet/roslyn/pull/44025#discussion_r446453378) [](ancestors = 446453378,446450749,446444174,446418702,446416546,446413877,446358005,446321935)

Reply:
> I can be convinced if you can come up with a test scenario that would demonstrate an incorrect behavior if the suggested changes were implemented.

If you mean to produce a methodimpl for covariant return scenarios when targeting runtimes that do not support covariant returns... that is simply invalid IL that doesn't obey the IL specification implemented by the runtime. Producing invalid IL is incorrect behavior.

---
In reply to: [446458615](https://github.com/dotnet/roslyn/pull/44025#discussion_r446458615) [](ancestors = 446458615,446453378,446450749,446444174,446418702,446416546,446413877,446358005,446321935)

Reply:
> If you mean to produce a methodimpl for covariant return scenarios when targeting runtimes that do not support covariant returns... that is simply invalid IL that doesn't obey the IL specification implemented by the runtime. Producing invalid IL is incorrect behavior.


I mean, please, give me something concrete. A test that fails with suggested changes, for example. So far you gave me just claims and guesses, nothing of substance. Please, be very specific, what is the scenario, what is the nature of failure, why do you think it is a failure, etc.
The change I suggest to make - before the return add a logic to check if we are dealing with covariant return case (i.e. there is a mismatch in return type, etc.), return true if that is the case.


---
In reply to: [446461000](https://github.com/dotnet/roslyn/pull/44025#discussion_r446461000) [](ancestors = 446461000,446458615,446453378,446450749,446444174,446418702,446416546,446413877,446358005,446321935)

Reply:
> See TestAmbiguousOverridesRefOut in CodeGenOverridingAndHiding as changed in iteration 33. Depending on the order of the declarations, the override would change from being correct to incorrect or vice versa (both tested).

Correct me if I am wrong, but I believe the test is not using covariant returns and, therefore, cannot be affected by suggested changes.

---
In reply to: [446463069](https://github.com/dotnet/roslyn/pull/44025#discussion_r446463069) [](ancestors = 446463069,446461000,446458615,446453378,446450749,446444174,446418702,446416546,446413877,446358005,446321935)

Reply:
> The compat council explicitly decided that we should not produce methodimpl entries for the ambiguous scenario on runtimes that are not known to have the fix to [https://github.com/dotnet/runtime/issues/38119](https://github.com/dotnet/runtime/issues/38119). The compat council's decision is documented in [https://github.com/dotnet/roslyn/issues/45453](https://github.com/dotnet/roslyn/issues/45453).

This decision doesn't apply to covariant return case because there is no backward compatibility scenario that involves covariant returns feature, compat council makes decisions about compatibility scenarios. Covariant Returns is a new feature, we never emitted anything like that before, with or without an override. And, I am pretty sure that with suggested changes, we wouldn't emit a MethodImpl for covariant return case in case of the ambiguity if runtime doesn't have the fix. If you disagree, all you have to do to prove me wrong is to provide a concrete test case failing with suggested changes.  

---
In reply to: [446471373](https://github.com/dotnet/roslyn/pull/44025#discussion_r446471373) [](ancestors = 446471373,446463069,446461000,446458615,446453378,446450749,446444174,446418702,446416546,446413877,446358005,446321935)

Reply:
> Perhaps this code from RetargetingMethodSymbol's base class should help:

Please take a look at this implementation of the API in SourceMemberMethodSymbol (note the use of this,  RequiresExplicitOverride, function):
```
        internal override bool IsMetadataNewSlot(bool ignoreInterfaceImplementationChanges = false)
        {
            if (IsExplicitInterfaceImplementation && _containingType.IsInterface)
            {
                // All implementations of methods from base interfaces should omit the newslot bit to ensure no new vtable slot is allocated.
                return false;
            }

            // If C# and the runtime don't agree on the overridden method,
            // then we will mark the method as newslot and specify the
            // override explicitly (see GetExplicitImplementationOverrides
            // in NamedTypeSymbolAdapter.cs).
            return this.IsOverride ?
                this.RequiresExplicitOverride() :
                this.IsMetadataVirtual(ignoreInterfaceImplementationChanges);
        }
```

For a method that overrides using covariant returns, this method is supposed to return true. When the overriding method is represented by a RetargetingMethodSymbol, IsMetadataNewSlot should also return true. Please add a targeted test to assert both expectations.

Here is another place (RetargetingMethodSymbol) where RequiresExplicitOverride makes a difference:
```
         internal MethodSymbol ExplicitlyOverriddenClassMethod
        {
            get
            {
                return
                    _underlyingMethod.RequiresExplicitOverride()
                        ? this.RetargetingTranslator.Retarget(_underlyingMethod.OverriddenMethod, MemberSignatureComparer.RetargetedExplicitImplementationComparer)
                        : null;
            }
        }
```

I strongly believe that the behavior of this function shouldn't depend on the fact whether the underlying compilation targets runtime with a fix or not. The behavior should be consistent across scenarios when there is an ambiguity at runtime and when there isn't an ambiguity at runtime. The reason is - there is no ambiguity at runtime for covariant return case, or the code simply would be illegal for that runtime. One might say that the mechanism we are using to detects the ambiguity has a flaw, it shouldn't be detecting any for a covariant return case. The proposed change basically mitigates that flaw. 


---
In reply to: [446472369](https://github.com/dotnet/roslyn/pull/44025#discussion_r446472369) [](ancestors = 446472369,446471373,446463069,446461000,446458615,446453378,446450749,446444174,446418702,446416546,446413877,446358005,446321935)